### PR TITLE
Add support for category option combo loop over

### PR DESCRIPTION
**Pull request by mestachs**
* https://github.com/BLSQ/orbf2/pull/145

### CODE REVIEW

File: lib/orbf/rules_engine/builders/activity_combo_variables_builder.rb [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+              lookup
+            ).convert(period))
+          end
+        end
+      end
+
+      def convert(period)
+        orgunits.each_with_object([]) do |orgunit, array|
+          package.all_activities_codes.each do |activity_code|
+            activity = package.activities.detect { |candidate| candidate.activity_code == activity_code }
```
Comment (pjaspers) [inline lib/orbf/rules_engine/builders/activity_combo_variables_builder.rb]: Couldn't we loop over package.activities?

File: lib/orbf/rules_engine/builders/activity_combo_variables_builder.rb [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+            activity.activity_states.each do |activity_state|
+              package.loop_over_combo[:category_option_combos].each do |category_option_combo|
+                category_option_combo_id = category_option_combo["id"] || category_option_combo[:id]
+                value = @lookup[[orgunit.ext_id, period, activity_state.ext_id, category_option_combo_id]]
+                value = (value.first[:value] || value.first["value"]) if value && value.first
```
Comment (pjaspers) [inline lib/orbf/rules_engine/builders/activity_combo_variables_builder.rb]: I'm kind of dreading the hash lookup with :value and "value", not sure if I'm dreading it enough to pull in HashWithIndifferentAccess though.

File: lib/orbf/rules_engine/builders/activity_formula_combo_variables_builder.rb [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        @orgunits = orgunits.out_list
+        @period = period
+        @tokens = {}
+        @period_facts = Orbf::RulesEngine::PeriodFacts.for(period)
+        @package_rule_codes = package.package_rules.flat_map(&:formulas).map(&:code).to_set
+        @activity_rule_codes = package.activity_rules.flat_map(&:formulas).map(&:code).to_set
+      end
+
+      def to_variables
+        return nil unless package.loop_over_combo
```
Comment (pjaspers) [inline lib/orbf/rules_engine/builders/activity_formula_combo_variables_builder.rb]: Are we expecting `nil` or `[]`?

File: lib/orbf/rules_engine/services/solver.rb [hunk 4]
```diff
@@ -14,6 +14,7 @@ def initialize(engine_version:)
       end
 
       def register_variables(vars)
+        return if vars.nil? || vars.empty?
```
Comment (pjaspers) [inline lib/orbf/rules_engine/services/solver.rb]: Ah we're setup to handle them both. `nil` or `empty`

File: spec/payers_spec.rb [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+  it "builds problem per category option combo" do
+    solved = build_and_solve(orgunits_full, dhis2_values)
+    problem = solved.solver.build_problem
+    puts JSON.pretty_generate(problem)
+
+    expect(problem["facility_spread_01_dhis2_payer_1_exportable_for_orgunit_id_and_2016q1"]).to eq(
+      "facility_spread_01_dhis2_payer_1_percentage_is_null_for_orgunit_id_and_2016q1 == 0"
+    )
+    expect(problem["facility_spread_01_dhis2_payer_1_percentage_calculated_for_orgunit_id_and_2016q1"]).to eq(
+      "(facility_spread_01_dhis2_payer_1_percentage_for_orgunit_id_and_2016q1/100) * 0.20"
```
Comment (pjaspers) [inline spec/payers_spec.rb]: Where is the '0.20' coming from?

File: spec/payers_spec.rb [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+    )
+    expect(problem["facility_spread_01_dhis2_payer_1_payer_payment_for_orgunit_id_and_2016q1"]).to eq(
+      "facility_spread_01_to_pay_for_orgunit_id_and_2016q1 * facility_spread_01_dhis2_payer_1_percentage_calculated_for_orgunit_id_and_2016q1"
+    )
+  end
+
+  it "should solve equations" do
+    solved = build_and_solve(orgunits_full, dhis2_values)
+
+    puts JSON.pretty_generate(solved.exported_values)
```
Comment (pjaspers) [inline spec/payers_spec.rb]: Can we hide this behind a verbose/debug flag? The less the tests output the more anxious I am when they do start outputting a lot.

File: spec/payers_spec.rb [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+          rules:                       [
+            Orbf::RulesEngine::Rule.new(
+              kind:     :activity,
+              formulas: [
+                build_activity_formula(
+                  "exportable", "percentage_is_null == 0",
+                  "should only export when percentage is filled"
+                ),
+                build_activity_formula(
+                  "percentage_calculated", " (percentage/100) * 0.20",
```
Comment (pjaspers) [inline spec/payers_spec.rb]: Found the '0.20' :)

File: lib/orbf/rules_engine/builders/activity_combo_variables_builder.rb (hunk 1 above)
Comment (mestachs) [reply lib/orbf/rules_engine/builders/activity_combo_variables_builder.rb]: copied the loop from activity_variables_builder don't really remember why I had to add the "uniqness by code"

File: spec/payers_spec.rb (hunk 6 above)
Comment (mestachs) [reply spec/payers_spec.rb]: replace with expect ;)

**Pull request merged**